Make spawn_n and map_n synchronous

FichteFoll opened the pull request:
These functions are specifically not blocking or awaiting, so they don't
need to be coroutines either.

Breaking change!

FichteFoll committed:
Make spawn_n and map_n synchronous

These functions are specifically not blocking or awaiting, so they don't
need to be coroutines either.

Breaking change!

gistart commented:
Hi. Thank you for your contribution!

Tests crash on this PR, for every python version:

```
==== short test summary info 
FAILED tests/test_callbacks.py::test_spawn_n - ZeroDivisionError: division by zero
FAILED tests/test_callbacks.py::test_map_n - TypeError: object list can't be used in 'await' expression
FAILED tests/test_callbacks.py::test_itermap[None] - TypeError: object list can't be used in 'await' expression
FAILED tests/test_callbacks.py::test_itermap[0.1] - TypeError: object list can't be used in 'await' expression
FAILED tests/test_callbacks.py::test_itermap[2] - TypeError: object list can't be used in 'await' expression
FAILED tests/test_map.py::test_itermap - TypeError: object list can't be used in 'await' expression
FAILED tests/test_spawn.py::test_spawns_behaviour - assert 5 == 0
==== 7 failed, 11 passed in 18.51s 
```

FichteFoll commented:
Working on it. FYI, I didn't get the tests to pass even on the old master.

FichteFoll committed:
asyncio.CancelledError does not subclass Exception

CancelledError does not subclass Exception anymore since 3.8
(and correctly so), but this broke control flow at a couple location
and was annoying to debug.

FichteFoll committed:
Enhance cancel test

Now verifies that coroutines can catch CancelledError and return a
value.

FichteFoll committed:
Fix usages of spawn_n and map_n in tests and readme

FichteFoll commented:
Fixed all tests (and refined the cancel test). There are quite a few warnings (DeprecationWarnings from `loop` parameters being removed in 3.10 and some instances of coroutines not being awaited), but I'll leave those for another time to resolve. I only ran the tests on 3.8.

FichteFoll commented:
Thinking about this again, I guess the `catch Exception` not catching CancelledErrors anymore might have been the main reason for my pain with that method, so I'm glad I took a look. I've had problems with the cancel method deadlocking but it wasn't clear to me how that could happen. Since one of the previously deadlocking tests now passes, I'm positive that I ran into the same issue.

gistart commented on asyncio_pool/base_pool.py:
```diff
@@ -106,7 +106,7 @@ async def _wrap(self, coro, future, cb=None, ctx=None):
         try:
             res = await coro
-        except Exception as _exc:
+        except BaseException as _exc:
```
This catch now includes stuff that, in my opinion, should never be catched in library code, e.g. `SystemExit` and `KeyboardInterrupt`.

https://docs.python.org/3/library/exceptions.html#exception-hierarchy

FichteFoll replied:
Yes but no. The job of these future wrappers is to specifically catch *all* exceptions raised in the wrapped future and then set them as the result future's exception. You're not swallowing the exception, you are making it available to be fetched on the call-site in the first palce. They would be swallowed if you didn't, since the exception would bubble up in the task and then never be retrieved.

gistart reviewed (changes requested):
I think `CancelledError` should be catched explicitly. In 3.8 it changed it's parent class from `Exception` to `BaseException`, but catching `BaseException` is not a good practice imo.
